src/Npgsql/Internal/NpgsqlConnector.cs

```diff
@@ -1752,7 +1752,11 @@ internal void PerformUserCancellation()
         // Wait before we've read all responses for the prepended queries
         // as we can't gracefully handle their cancellation.
         // Break makes sure that it's going to be set even if we fail while reading them.
-        ReadingPrependedMessagesMRE.Wait();
+
+        // We don't wait indefinitely to avoid deadlocks from synchronous CancellationToken.Register
+        // See #5032
+        if (!ReadingPrependedMessagesMRE.Wait(0))
+            return;
```

Comment:
Exit this Monitor before returning.
https://github.com/npgsql/npgsql/blob/9cc91142c876590bc11106bd1d3df5ee4433cb7b/src/Npgsql/Internal/NpgsqlConnector.cs#L1749

Reply:
Indeed, you're right, thanks! I'll just put it below into `try/finally` block.